### PR TITLE
[executor] handling read/allocate/write-keys in the execution engine

### LINKED ISSUE #701: [executor] parallelizing reads in the execution engine
See below

Prereq: #693 
Related: #709

### PR DESCRIPTION
**Pull request by wlawt**
Closes #701, #709

### Overview

Please see the 2nd comment in #701 for the overall approach. Some modifications were made from the original idea. That being, in order to ensure reads can't happen before writing, and writing can't happen before reading, the executor keeps track of the last blocked task in `edges` known as the `Parent` within `KeyData`. This allows us to execute the tasks in the order in which they were enqueued, in the case of high contention.

### Test Plan

- Added unit tests in `executor_test.go`, coverage is ~97.4%
- Existing unit tests in `executor_test.go` pass
- Integration tests in MorpheusVM and TokenVM pass
 
### TODOs

- [x] Add more unit tests
- [ ] Handle GetValue caching

### CODE REVIEW

File: executor/executor.go
Comment (wlawt) [inline executor/executor.go:40]: nit: is there a clearer name

File: chain/processor.go
Comment (patrick-ogrady) [inline chain/processor.go:95]: We should avoid holding the lock while fetching keys from the database imo

File: chain/processor.go
Comment (patrick-ogrady) [inline chain/processor.go:110]: We also update the cache with an `RLock` (which probably is unexpected behavior)

File: chain/processor.go
Comment (patrick-ogrady) [inline chain/processor.go:113]: Even if we did want to hold the lock during the entirety of this disk fetch, we wouldn't prevent multiple fetches of the same key because we haven't updated `cache` by the time this lock is released.

File: executor/executor.go
Comment (patrick-ogrady) [inline executor/executor.go:158]: I think this should be `Write` or `Allocate`?

File: executor/executor.go
Comment (patrick-ogrady) [inline executor/executor.go:160]: I don't think it matters what permissions the dependencies have, we need to wait?

Or is the idea that we are "future-looking" and being explicit with our checks?

File: executor/executor.go
Comment (patrick-ogrady) [inline executor/executor.go:173]: In the previous code, we didn't run a transaction if there were dependencies. I'm not sure what `e.blocking` adds?

File: executor/executor.go
Comment (patrick-ogrady) [inline executor/executor.go:155]: I think you could `continue` inside of this `if`?

File: executor/executor.go
Comment (patrick-ogrady) [inline executor/executor.go:149]: You need to register the type of blocker this is `state.Read` so that when complete, `lt` can potentially unblock multiple ids.

Comment (patrick-ogrady) [inline executor/executor.go:149]: If this is a read dependency, we don't necessarily just want to register on the preceding tx...we would want to register on the last blocked transaction (otherwise non-blocking txs will be unnecessarily serialized).

Comment (patrick-ogrady) [reply executor/executor.go:149]: This may also cause a bug if the `lt` is blocked on a write but we don't add a dependency because `lt`'s usage of a key is a read.

File: executor/executor.go
Comment (patrick-ogrady) [inline executor/executor.go:153]: As mentioned below, I don't know what `e.blocking` adds and think it should be removed.

Blocking is implied via `lt.blocking`. No transaction will be executed until all blockers are cleared by its dependendencies.

File: executor/executor.go
Comment (patrick-ogrady) [reply executor/executor.go:149]: This could be fixed by adding `KeyData.Parent` and we iterate backwards until we get to executed or a write/alloc I think.

File: chain/processor.go
Comment (patrick-ogrady) [inline chain/processor.go:83]: I'm not sure a `RLock` helps prevent concurrent reads here as any number of transactions would see the value is missing in the cache and try to read it?

File: executor/executor.go
Comment (patrick-ogrady) [inline executor/executor.go:157]: Haven't reviewed the logic, but you should do a `continue` after this line instead of doing a monster `if/else`

Comment (wlawt) [reply executor/executor.go:157]: oops forgot, updated now

File: executor/executor.go
Comment (patrick-ogrady) [inline executor/executor.go:157]: There is no `Parent` here?

File: executor/executor.go
Comment (patrick-ogrady) [inline executor/executor.go:216]: With this approach, you'll want to remove `ids` from `ConcurrentReads` when they finish execution? Seems like a way to avoid a lot of extra iteration here (and complexity).

File: executor/executor.go
Comment (patrick-ogrady) [inline executor/executor.go:224]: Why would you want to clear `ConcurrentReads` here? Seems like it should be included in the `if` above?

File: executor/executor.go
Comment (patrick-ogrady) [inline executor/executor.go:154]: This is a tricky condition to consider. I agree it helps with tracking dependencies.

File: executor/executor.go
Comment (patrick-ogrady) [inline executor/executor.go:148]: I don't love `parent` here given this `keyData` can be either the `parent` or some value back in the ancestry of keys.

File: executor/executor.go
Comment (patrick-ogrady) [inline executor/executor.go:160]: I think `e.tasks[key.ID]` is much easier to reason about.

File: executor/executor.go
Comment (patrick-ogrady) [inline executor/executor.go:55]: should be private struct `keyData`

File: executor/executor.go
Comment (patrick-ogrady) [inline executor/executor.go:157]: Don't love `Parent` here. Not sure what to replace it with but it is really `LastWriteAllocateOrRead` lol

File: executor/executor.go
Comment (patrick-ogrady) [reply executor/executor.go:148]: Maybe just `key`?

File: executor/executor.go
Comment (patrick-ogrady) [reply executor/executor.go:157]: Maybe just `ID`?

File: executor/executor.go
Comment (patrick-ogrady) [inline executor/executor.go:170]: Could this lead to the same `id` being added to `pt.blocking` twice?

If `parent.Permissions == state.Write` and 1 tx comes in with `v==state.Read`, it will add here. If another tx comes in with `v==state.Write`, then it will iterate through the concurrent reads below and add the item to `pt.blocking`.

File: executor/executor.go
Comment (patrick-ogrady) [reply executor/executor.go:216]: I would think `if executed`, you should assume there is nothing to do here besides update `ID` and `Permissions`?

**Review by patrick-ogrady — CHANGES_REQUESTED** [review]

File: executor/executor.go
Comment (wlawt) [reply executor/executor.go]: reposting for visibility:

if i'm understanding this comment correctly, suppose we have the following 3 transactions:
T1 W
T2 R
T3 W

here, in T1 W, it's "parent" is T1. suppose it doesn't yet execute, for some reason

then in T2 R, it gets blocked by its "parent" which is T1, so we have the following:
T1.blocking = (T2)
T2.dep = (T1)
concurrentReads = (T2)

then when we run T3 W, we will go through concurrentReads and update the following:
T2.blocking = (T3)
T3.dep = (T2)

then we clear concurrentReads, and then finally, we update the "parent" to be T3

so in this walkthrough we don't add to pt.blocking twice. if a key has Allocate/Write, it will always be updated to itself

File: executor/executor.go
Comment (wlawt) [reply executor/executor.go]: reposting for visibility:

we want to clear concurrentReads here before updating because if we had the case of only reads like: R->R->R->R, and the first R already executed, concurrentReads was populated with that first R, but was already done by the time we reached this part of the code, so we need to reset it to avoid adding unwanted concurrentReads

**Comment by wlawt** [comment]
Please see #779 for the continuation of the PR (the branch is no longer a fork of the repo)

**Pull request closed without merging**
